### ISSUE TITLE
Overlapping routes?

### ISSUE THREAD

**Issue by dallonf** [Netflix/falcor#456]
tl;dr: When multiple routes match the same path, will the Router optimize by choosing the route that matches the most paths in the current request?

tl:

So let's say I'm building a Router based on an existing REST API:

```
// GET /api/people
{
  data: [{
    id: 1,
    name: "Bruce Wayne"
  }]
}
```

And, of course, `/api/people/1` would return just Wayne Enterprises.

It would make sense to model this in a JSON Graph something like this (technically the array should also have a length, but I'm ignoring that for simplicity):

```
{
  personList: {
    0: $ref(['peopleById', 1])
  },
  peopleById: {
    1: {
      name: "Bruce Wayne"
    }
  }
}
```

According to the official sample, I would build two routes for this:
- `personList[{integers}]` (calls `GET /api/people`)
- `peopleById[{integers}].name` (calls `GET /api/people/:id`)

But if I want to request `personList[0].name`, I wind up making two requests to the backing REST API, even though all the data I needed was available in one request. So really, what I would want is a route list like this:
- `personList[{integers}]` (calls `GET /api/people`)
- `personList[{integers}].name` (calls `GET /api/people`)
- `peopleById[{integers}].name` (calls `GET /api/people/:id`)

If I request `personList[0].name`, I'd expect it to use just the second route, because it _most specifically_ fulfills the requested path. (It's not currently documented what would happen if I tried this in the current version)

But there's something weird happening here, too! The second route is providing values _through_ a ref. Is that even possible? _Should_ that be possible?

And then there's the whole matter of values potentially being fulfilled by more than one route, which could result in some really _weird_ non-deterministic behavior if you're not very careful.

At the same time, there probably should be _some_ way to efficiently handle cases like this, where one service call can provide many values in various places on the graph. Right now, it seems to primarily handle the opposite case, where various service calls are needed to fulfill a local cluster of the graph.

**Comment by dallonf** [comment]
After a few minutes of posting this, I realized it might actually be better for routes to _communicate_ with each other somehow and allow them shortcut a fetch, rather than overlapping.

So let's take the request for `personList[0].name` again... it would first hit the `personList[{integers}]` route as always, which would make a request to `GET /api/people` and get the `ref` to `peopleById[0]`. It would _also_ then populate some sort of cache with the contents of the array.

Then, when the router goes to `peopleById[{integers}].name`, the route would _first_ check to see if any data is available already in the aforementioned cache, and then perform a `GET /api/people/1` request otherwise.

This actually wouldn't be terribly difficult given the current Router API (just put the array element in `this.peopleCache[0]` or something), but given that caching is already one of the core concerns of Falcor, it might make sense to have a first-class feature, or at least document a preferred technique.

**Comment by jhusain** [comment]
You hit on the solution to the problem. You need to introduce a cache to avoid making sequential requests for the same information. Together batching and caching are relatively simple solutions that can lead to surprisingly efficient data access patterns without the need to couple your routes to each other. This is how Netflix interacts with the back end from its Router.

As for introducing support for this at the Router level, it's probably more flexible to build generic service layer helpers that work by convention and mix caching and batching into your services get[entity]ById and so forth. A few early adopters have done just that. Be interesting to release them. @bcardi is using something like this.

**Comment by dallonf** [comment]
Thanks for the reply! One more question about prefetching... let's say that `/people` has a one-to-many relationship called "sidekicks", accessible ordinarily from `GET /api/people/1/sidekicks`, but also can be inlined in another request:

```
GET /api/people?includeSidekicks=1
{
  data: [{
    id: 1,
    name: "Bruce Wayne",
    sidekicks: [{
      id: 2,
      name: "Dick Grayson"
    }]
  }]
}
```

If I were to request `peopleList[0].sidekicks[0]`, it would make sense to include that parameter in the query. In order to do that, though, I would either need to be able to specify a more specific query like `peopleList[{integers}].sidekicks[{integers}].name` (but that way lies madness), or the `peopleList[{integers}]` route would need to have information about future paths or PathSets in the same batched request that have yet to be resolved. Again, not outside the realm of possibility and I'd imagine the API being fairly simple. It might even already be possible, although there's no API docs for Router yet and I don't have time right now to read the source.

Also, I noticed this same issue being discussed in the Router repo itself (https://github.com/Netflix/falcor-router/issues/79) - should I migrate over to that thread?

**Comment by jhusain** [comment]
If you're underlying service layer supports it, you should make the most coarse-grained request to the backend you possibly can. That is why the Router supports higher precedence matching for longer paths.

At the moment this is your best bet:

Match the more specific path, and the various other paths:
```
peopleList[{integers}]
personsById[{integers}].sidekicks[{integers}]
sidekicksById[{integers}][...]
peopleList[{integers}].sidekicks[{integers}] // remember this just returns a reference to sidekicksById
```

Both the first and last routes will call the same service presumably. In the service for retrieving the people list, you can track whether at least one request has come in for sidekicks. Just batch schedule a service that wraps the REST API like in this example...

https://github.com/Netflix/falcor-router-demo/blob/master/recommendation-service.js#L13

... and set a flag on the batch if at least one request comes in for sidekicks. This way of two parallel requests come in for peopleList[5] and peopleList[5].sidekicks[7]

We are currently evaluating different ways of allowing you to serve multiple routes with access to the rest of the path for context.

**Comment by ktrott** [comment]
@dallonf Are your questions answered? May we close this issue?

@jhusain Are there any gaps in documentation or examples that we could fill to capture this for the next person? Should we add a Q&A section that captures any of this?

**Comment by dallonf** [comment]
I think this answers my questions , yes.
